File: sdk/node/src/resources.ts
import { BaseClient } from './base.js'
import type {
  Approval,
  Domain,
  DomainSummary,
  EvalCase,
  EvalRun,
  EvalSet,
  EvalSetSummary,
  ListEnvelope,
  Session,
  SessionSummary,
  Trace,
  TraceSummary,
} from './types.js'

export interface ListDomainsParams {
  limit?: number
  offset?: number
}

export class DomainRegistryClient extends BaseClient {
  list(params: ListDomainsParams = {}): Promise<ListEnvelope<DomainSummary>> {
    return this._get<ListEnvelope<DomainSummary>>(`/domains${this._queryString(params)}`)
  }

  get(id: string): Promise<Domain> {
    return this._get<Domain>(`/domains/${encodeURIComponent(id)}`)
  }

  getYaml(id: string): Promise<string> {
    return this._request<string>('GET', `/domains/${encodeURIComponent(id)}/yaml`, undefined, {
      Accept: 'application/yaml',
    })
  }

  registerYaml(yaml: string): Promise<Domain> {
    return this._request<Domain>('POST', '/domains', yaml, {
      'Content-Type': 'application/x-yaml',
    })
  }

  updateYaml(id: string, yaml: string): Promise<Domain> {
    return this._request<Domain>('PUT', `/domains/${encodeURIComponent(id)}`, yaml, {
      'Content-Type': 'application/x-yaml',
    })
  }

  remove(id: string): Promise<void> {
    return this._delete<void>(`/domains/${encodeURIComponent(id)}`)
  }

  validateYaml(id: string, yaml: string): Promise<{ valid: boolean; errors?: string[] }> {
    return this._request<{ valid: boolean; errors?: string[] }>(
      'POST',
      `/domains/${encodeURIComponent(id)}/validate`,
      yaml,
      { 'Content-Type': 'application/x-yaml' },
    )
  }
}

export interface TriggerSessionOptions {
  domainId: string
  trigger?: Record<string, unknown>
}

export interface ListSessionsParams {
  domain?: string
  state?: string
  limit?: number
  offset?: number
}

export class SessionsClient extends BaseClient {
  list(params: ListSessionsParams = {}): Promise<ListEnvelope<SessionSummary>> {
    return this._get<ListEnvelope<SessionSummary>>(`/sessions${this._queryString(params)}`)
  }

  get(id: string): Promise<Session> {
    return this._get<Session>(`/sessions/${encodeURIComponent(id)}`)
  }

  trigger(options: TriggerSessionOptions): Promise<Session> {
    return this._post<Session>('/sessions', {
      domain_id: options.domainId,
      trigger: options.trigger ?? {},
    })
  }

  cancel(id: string): Promise<Session> {
    return this._post<Session>(`/sessions/${encodeURIComponent(id)}/cancel`)
  }

  rerun(id: string): Promise<Session> {
    return this._post<Session>(`/sessions/${encodeURIComponent(id)}/rerun`)
  }
}

export interface ListTracesParams {
  domain?: string
  session?: string
  agent?: string
  from?: string
  to?: string
  limit?: number
  offset?: number
}

export class TracesClient extends BaseClient {
  list(params: ListTracesParams = {}): Promise<ListEnvelope<TraceSummary>> {
    return this._get<ListEnvelope<TraceSummary>>(`/traces${this._queryString(params)}`)
  }

  get(traceId: string): Promise<Trace> {
    return this._get<Trace>(`/traces/${encodeURIComponent(traceId)}`)
  }
}

export interface ListApprovalsParams {
  domain?: string
  session?: string
  status?: string
}

export class ApprovalsClient extends BaseClient {
  list(params: ListApprovalsParams = {}): Promise<ListEnvelope<Approval>> {
    return this._get<ListEnvelope<Approval>>(`/approvals${this._queryString(params)}`)
  }

  get(id: string): Promise<Approval> {
    return this._get<Approval>(`/approvals/${encodeURIComponent(id)}`)
  }

  approve(id: string): Promise<Approval> {
    return this._post<Approval>(`/approvals/${encodeURIComponent(id)}/approve`)
  }

  reject(id: string): Promise<Approval> {
    return this._post<Approval>(`/approvals/${encodeURIComponent(id)}/reject`)
  }
}

export interface CreateEvalSetOptions {
  setId: string
  domainId: string
  description?: string
  cases: EvalCase[]
}

export class EvalsClient extends BaseClient {
  listSets(): Promise<ListEnvelope<EvalSetSummary>> {
    return this._get<ListEnvelope<EvalSetSummary>>('/evals')
  }

  getSet(id: string): Promise<EvalSet> {
    return this._get<EvalSet>(`/evals/${encodeURIComponent(id)}`)
  }

  createSet(options: CreateEvalSetOptions): Promise<EvalSet> {
    return this._post<EvalSet>('/evals', {
      set_id: options.setId,
      domain_id: options.domainId,
      description: options.description,
      cases: options.cases,
    })
  }

  updateSet(
    id: string,
    options: { description?: string; cases: EvalCase[] },
  ): Promise<EvalSet> {
    return this._put<EvalSet>(`/evals/${encodeURIComponent(id)}`, {
      description: options.description,
      cases: options.cases,
    })
  }

  removeSet(id: string): Promise<void> {
    return this._delete<void>(`/evals/${encodeURIComponent(id)}`)
  }

  runSet(id: string): Promise<{ run_id: string; state: string }> {
    return this._post<{ run_id: string; state: string }>(`/evals/${encodeURIComponent(id)}/run`)
  }

  listRuns(setId: string): Promise<ListEnvelope<EvalRun>> {
    return this._get<ListEnvelope<EvalRun>>(`/evals/${encodeURIComponent(setId)}/runs`)
  }

  getRun(setId: string, runId: string): Promise<EvalRun> {
    return this._get<EvalRun>(
      `/evals/${encodeURIComponent(setId)}/runs/${encodeURIComponent(runId)}`,
    )
  }
}

File: sdk/go/resources.go
package hnsx

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
)

// DomainClient operates on domains.
type DomainClient struct {
	client *Client
}

// List returns all registered domains.
func (c *DomainClient) List(ctx context.Context, limit, offset int) (*ListEnvelope[DomainSummary], error) {
	path := "/domains" + queryString(map[string]any{"limit": limit, "offset": offset})
	data, err := c.client.doJSON(ctx, "GET", path, nil)
	if err != nil {
		return nil, err
	}
	var out ListEnvelope[DomainSummary]
	if err := json.Unmarshal(data, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Get returns a single domain by ID.
func (c *DomainClient) Get(ctx context.Context, id string) (*Domain, error) {
	data, err := c.client.doJSON(ctx, "GET", fmt.Sprintf("/domains/%s", id), nil)
	if err != nil {
		return nil, err
	}
	var out Domain
	if err := json.Unmarshal(data, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// RegisterYAML uploads a domain YAML and returns the registered domain.
func (c *DomainClient) RegisterYAML(ctx context.Context, yaml string) (*Domain, error) {
	resp, err := c.client.request(ctx, "POST", "/domains", bytes.NewBufferString(yaml), map[string]string{"Content-Type": "application/x-yaml"})
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	data, err := decodeResponse(resp)
	if err != nil {
		return nil, err
	}
	var out Domain
	if err := json.Unmarshal(data, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// SessionClient operates on sessions.
type SessionClient struct {
	client *Client
}

// List returns sessions.
func (c *SessionClient) List(ctx context.Context, domain, state string, limit, offset int) (*ListEnvelope[SessionSummary], error) {
	path := "/sessions" + queryString(map[string]any{"domain": domain, "state": state, "limit": limit, "offset": offset})
	data, err := c.client.doJSON(ctx, "GET", path, nil)
	if err != nil {
		return nil, err
	}
	var out ListEnvelope[SessionSummary]
	if err := json.Unmarshal(data, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Get returns a single session by ID.
func (c *SessionClient) Get(ctx context.Context, id string) (*Session, error) {
	data, err := c.client.doJSON(ctx, "GET", fmt.Sprintf("/sessions/%s", id), nil)
	if err != nil {
		return nil, err
	}
	var out Session
	if err := json.Unmarshal(data, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Trigger starts a new session for a domain.
func (c *SessionClient) Trigger(ctx context.Context, domainID string, trigger map[string]any) (*Session, error) {
	data, err := c.client.doJSON(ctx, "POST", "/sessions", map[string]any{"domain_id": domainID, "trigger": trigger})
	if err != nil {
		return nil, err
	}
	var out Session
	if err := json.Unmarshal(data, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Cancel cancels a running session.
func (c *SessionClient) Cancel(ctx context.Context, id string) (*Session, error) {
	data, err := c.client.doJSON(ctx, "POST", fmt.Sprintf("/sessions/%s/cancel", id), nil)
	if err != nil {
		return nil, err
	}
	var out Session
	if err := json.Unmarshal(data, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// TraceClient operates on traces.
type TraceClient struct {
	client *Client
}

// List returns traces.
func (c *TraceClient) List(ctx context.Context, domain, sessionID string, limit, offset int) (*ListEnvelope[TraceSummary], error) {
	path := "/traces" + queryString(map[string]any{"domain": domain, "session": sessionID, "limit": limit, "offset": offset})
	data, err := c.client.doJSON(ctx, "GET", path, nil)
	if err != nil {
		return nil, err
	}
	var out ListEnvelope[TraceSummary]
	if err := json.Unmarshal(data, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Get returns a single trace by ID.
func (c *TraceClient) Get(ctx context.Context, traceID string) (*Trace, error) {
	data, err := c.client.doJSON(ctx, "GET", fmt.Sprintf("/traces/%s", traceID), nil)
	if err != nil {
		return nil, err
	}
	var out Trace
	if err := json.Unmarshal(data, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// ApprovalClient operates on approvals.
type ApprovalClient struct {
	client *Client
}

// List returns approvals.
func (c *ApprovalClient) List(ctx context.Context, domain, sessionID, status string) (*ListEnvelope[Approval], error) {
	path := "/approvals" + queryString(map[string]any{"domain": domain, "session": sessionID, "status": status})
	data, err := c.client.doJSON(ctx, "GET", path, nil)
	if err != nil {
		return nil, err
	}
	var out ListEnvelope[Approval]
	if err := json.Unmarshal(data, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Approve approves a pending approval.
func (c *ApprovalClient) Approve(ctx context.Context, id string) (*Approval, error) {
	data, err := c.client.doJSON(ctx, "POST", fmt.Sprintf("/approvals/%s/approve", id), nil)
	if err != nil {
		return nil, err
	}
	var out Approval
	if err := json.Unmarshal(data, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Reject rejects a pending approval.
func (c *ApprovalClient) Reject(ctx context.Context, id string) (*Approval, error) {
	data, err := c.client.doJSON(ctx, "POST", fmt.Sprintf("/approvals/%s/reject", id), nil)
	if err != nil {
		return nil, err
	}
	var out Approval
	if err := json.Unmarshal(data, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// EvalClient operates on eval sets and runs.
type EvalClient struct {
	client *Client
}

// ListSets returns all eval sets.
func (c *EvalClient) ListSets(ctx context.Context) (*ListEnvelope[EvalSetSummary], error) {
	data, err := c.client.doJSON(ctx, "GET", "/evals", nil)
	if err != nil {
		return nil, err
	}
	var out ListEnvelope[EvalSetSummary]
	if err := json.Unmarshal(data, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// GetSet returns an eval set by ID.
func (c *EvalClient) GetSet(ctx context.Context, id string) (*EvalSet, error) {
	data, err := c.client.doJSON(ctx, "GET", fmt.Sprintf("/evals/%s", id), nil)
	if err != nil {
		return nil, err
	}
	var out EvalSet
	if err := json.Unmarshal(data, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// CreateSet creates a new eval set.
func (c *EvalClient) CreateSet(ctx context.Context, setID, domainID, description string, cases []EvalCase) (*EvalSet, error) {
	payload := map[string]any{"set_id": setID, "domain_id": domainID, "description": description, "cases": cases}
	data, err := c.client.doJSON(ctx, "POST", "/evals", payload)
	if err != nil {
		return nil, err
	}
	var out EvalSet
	if err := json.Unmarshal(data, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// RunSet starts an eval run.
func (c *EvalClient) RunSet(ctx context.Context, id string) (map[string]any, error) {
	data, err := c.client.doJSON(ctx, "POST", fmt.Sprintf("/evals/%s/run", id), nil)
	if err != nil {
		return nil, err
	}
	var out map[string]any
	if err := json.Unmarshal(data, &out); err != nil {
		return nil, err
	}
	return out, nil
}
